moveit_fake_controller_manager/src/moveit_fake_controllers.cpp

```diff
@@ -0,0 +1,288 @@
+  return true;
+}
+
+bool ThreadedController::cancelExecution()
+{
+  cancelTrajectory();
+  ROS_INFO("Fake trajectory execution cancelled");
+  status_ = moveit_controller_manager::ExecutionStatus::ABORTED;
+  return true;
+}
+
+bool ThreadedController::waitForExecution(const ros::Duration &)
+{
+  thread_.join();
+  status_ = moveit_controller_manager::ExecutionStatus::SUCCEEDED;
+  return true;
+}
+
+moveit_controller_manager::ExecutionStatus ThreadedController::getLastExecutionStatus()
+{
+  return status_;
+}
+
+
+ViaPointController::ViaPointController(const std::string &name,
+                                       const std::vector<std::string> &joints,
+                                       const ros::Publisher &pub) :
+  ThreadedController(name, joints, pub)
+{
+}
+
+ViaPointController::~ViaPointController()
+{
+}
+
+void ViaPointController::execTrajectory(const moveit_msgs::RobotTrajectory &t)
+{
+  ROS_INFO("Fake execution of trajectory");
+  sensor_msgs::JointState js;
+  js.header = t.joint_trajectory.header;
+  js.name = t.joint_trajectory.joint_names;
+
+  // publish joint states for all intermediate via points of the trajectory
+  // no further interpolation
+  ros::Time startTime = ros::Time::now();
+  for (std::vector<trajectory_msgs::JointTrajectoryPoint>::const_iterator
+       via = t.joint_trajectory.points.begin(), end = t.joint_trajectory.points.end(); !cancelled() && via != end; ++via) {
+    js.position = via->positions;
+    js.velocity = via->velocities;
+    js.effort = via->effort;
+
+    ros::Duration waitTime = via->time_from_start - (ros::Time::now() - startTime);
+    if (waitTime.toSec() > 1e-3) {
+      ROS_DEBUG("Fake execution: waiting %0.1fs for next via point, %ld remaining", waitTime.toSec(), end-via);
+      waitTime.sleep();
+    }
+    js.header.stamp = ros::Time::now();
+    pub_.publish(js);
+  }
+  ROS_DEBUG("Fake execution of trajectory: done");
+}
+
+
+InterpolatingController::InterpolatingController(const std::string &name, const std::vector<std::string> &joints, const ros::Publisher &pub)
+  : ThreadedController(name, joints, pub)
+  , rate_(10)
+{
+  double r;
+  if (ros::param::get("~rate", r))
+    rate_ = ros::WallRate(r);
+}
+
+InterpolatingController::~InterpolatingController()
+{
+}
+
+namespace
+{
+void interpolate(sensor_msgs::JointState &js,
+                 const trajectory_msgs::JointTrajectoryPoint &prev,
+                 const trajectory_msgs::JointTrajectoryPoint &next,
+                 const ros::Duration &elapsed)
+{
+  double duration = (next.time_from_start - prev.time_from_start).toSec();
+  double alpha = duration > std::numeric_limits<double>::epsilon() ? (elapsed - prev.time_from_start).toSec() / duration
+                                                                   : 1.0;
+  js.position.resize(prev.positions.size());
+  for (std::size_t i = 0, end = prev.positions.size(); i < end; ++i)
```

Comment:
interesting way of doing a for loop - does the ``end`` var actually have better performance? seems like a really unusual way to do things